GearRent/Networking/APIManager.m

```diff
@@ -0,0 +1,144 @@
+#import "Edge.h"
+
+@implementation APIManager
+
++ (void)fetchListingsWithCoordinates:(NSArray *)coordinates completion:(void(^)(NSArray *listings, NSError *error))completion{
+    // Find all the geohashes within the polygon and return nil if no geohashes with precision 7 exist
+    NSMutableSet *geohashesWithinPolygon = [[NSMutableSet alloc] init];
+    for(int i = 1; i <= 7; i ++){
+        geohashesWithinPolygon = [self findAllGeohashesWithinPolygon:coordinates precision:i];
+        if([geohashesWithinPolygon count] > 0){
+            break;
+        }
+    }
+    if([geohashesWithinPolygon count] == 0){
+        NSError *error = [NSError errorWithDomain:@"someDomain" code:404 userInfo:@{@"Error reason": @"Polygon too small"}];
+        completion(nil, error);
+    }
+    [PFCloud callFunctionInBackground:@"getListingsWithGeohashes"
+                       withParameters:@{@"geohashes": [geohashesWithinPolygon allObjects]}
+                                block:^(NSArray *listings, NSError *error) {
+      if (!error) {
+          NSLog(@"%@", listings); // complete with listings after
+      }
+    }];
+}
+
++ (NSMutableSet *)findAllGeohashesWithinPolygon:(NSArray *)coordinates precision:(int)precision {
+    NSMutableSet *uncheckedGeohashes = [[NSMutableSet alloc] init];
+    NSMutableSet *insideGeohashes = [[NSMutableSet alloc] init];
+    NSMutableSet *outsideGeohashes = [[NSMutableSet alloc] init];
+    for(int i = 0; i < coordinates.count; i++){
+        CLLocation *location = (CLLocation *) coordinates[i];
+        GNGeoHash *gh = [GNGeoHash withCharacterPrecision: location.coordinate.latitude andLongitude:location.coordinate.longitude andNumberOfCharacters:precision];
+        [uncheckedGeohashes addObject:gh];
+    }
+    while (uncheckedGeohashes.count > 0) {
+        GNGeoHash *currGH = (GNGeoHash *)[uncheckedGeohashes allObjects][0];
+        if([self polygonContainsGeohash:coordinates geohash:currGH]){
+            [insideGeohashes addObject:currGH];
+            NSArray *neighbors = [currGH getAdjacent];
+            for(GNGeoHash *neighbor in neighbors){
+                if(![insideGeohashes containsObject:neighbor] && ![outsideGeohashes containsObject:neighbor] && ![uncheckedGeohashes containsObject:neighbor]){
+                    [uncheckedGeohashes addObject:neighbor];
+                }
+            }
+        } else{
+            [outsideGeohashes addObject:currGH];
+        }
+    }
+    return insideGeohashes;
+}
+
++ (BOOL)polygonContainsGeohash:(NSArray *)coordinates geohash:(GNGeoHash *)geohash {
+    NSMutableArray *currGHCorners = [self getFourConersOfGeohash:geohash];
+    for(int i = 0; i < currGHCorners.count; i ++){
+        CLLocation *currCorner = currGHCorners[i];
+        if(![self polygonContainsPoint:coordinates point:currCorner]){
+            return NO;
+        }
+    }
+    return YES;
+}
+
++ (BOOL)polygonContainsPoint:(NSArray *)coordinates point:(CLLocation *)point {
```

Comment:
nit: might be easier to read to rename `coordinates` into `polygonCoordinates`

Reply:
Good idea